File: nestjs/src/cases/case2-queue/queue.service.ts
import { Injectable } from '@nestjs/common';
import { randomUUID } from 'crypto';
import { RedisService } from '../../core/redis/redis.service';

export const STREAM_NAME = 'stream:jobs';
export const CONSUMER_GROUP = 'group:workers';

@Injectable()
export class QueueService {
  constructor(private readonly redisService: RedisService) {}

  private get redis() {
    return this.redisService.getClient();
  }

  async ensureConsumerGroup(): Promise<void> {
    try {
      if (typeof (this.redis as any).xgroup === 'function') {
        await (this.redis as any).xgroup('CREATE', STREAM_NAME, CONSUMER_GROUP, '0', 'MKSTREAM');
      }
    } catch (e: any) {
      // BUSYGROUP or unsupported in mock
    }
  }

  async enqueueJob(taskType: string, payload: any): Promise<string> {
    await this.ensureConsumerGroup();
    const jobId = randomUUID();
    const jobData: Record<string, string> = {
      job_id: jobId,
      task_type: taskType,
      payload: JSON.stringify(payload),
      status: 'PENDING',
      result: '',
      created_at: (Date.now() / 1000).toString(),
      updated_at: (Date.now() / 1000).toString(),
      worker_id: '',
    };

    await this.redis.hset(`job:${jobId}`, jobData);
    try {
      if (typeof (this.redis as any).xadd === 'function') {
        await (this.redis as any).xadd(STREAM_NAME, '*', 'job_id', jobId, 'task_type', taskType);
      }
    } catch {
      // xadd unsupported in mock environment
    }
    return jobId;
  }

  async getJob(jobId: string): Promise<any | null> {
    const data = await this.redis.hgetall(`job:${jobId}`);
    if (!data || Object.keys(data).length === 0) {
      return null;
    }
    try {
      data.payload = JSON.parse(data.payload);
    } catch {
      // Keep as string
    }
    return data;
  }

  async updateJobStatus(jobId: string, status: string, workerId = '', result = ''): Promise<void> {
    const updates: Record<string, string> = {
      status,
      updated_at: (Date.now() / 1000).toString(),
    };
    if (workerId) updates.worker_id = workerId;
    if (result) updates.result = result;
    await this.redis.hset(`job:${jobId}`, updates);
  }
}

File: fastapi/app/cases/case2_queue/queue_service.py
import json
import time
import uuid
from typing import Any
from redis.asyncio import Redis

STREAM_NAME = "stream:jobs"
CONSUMER_GROUP = "group:workers"


class JobQueueService:
    def __init__(self, redis: Redis):
        self.redis = redis

    async def ensure_consumer_group(self) -> None:
        try:
            await self.redis.xgroup_create(
                name=STREAM_NAME,
                groupname=CONSUMER_GROUP,
                id="0",
                mkstream=True,
            )
        except Exception as e:
            # Group already exists
            if "BUSYGROUP" not in str(e):
                raise

    async def enqueue_job(self, task_type: str, payload: dict[str, Any]) -> str:
        await self.ensure_consumer_group()
        job_id = str(uuid.uuid4())
        job_data = {
            "job_id": job_id,
            "task_type": task_type,
            "payload": json.dumps(payload),
            "status": "PENDING",
            "result": "",
            "created_at": str(time.time()),
            "updated_at": str(time.time()),
            "worker_id": "",
        }
        # Save job state in Redis Hash
        await self.redis.hset(f"job:{job_id}", mapping=job_data)
        # Push message into Redis Stream
        await self.redis.xadd(
            STREAM_NAME,
            {"job_id": job_id, "task_type": task_type},
        )
        return job_id

    async def get_job(self, job_id: str) -> dict[str, Any] | None:
        data = await self.redis.hgetall(f"job:{job_id}")
        if not data:
            return None
        if "payload" in data and isinstance(data["payload"], str):
            try:
                data["payload"] = json.loads(data["payload"])
            except Exception:
                pass
        return data

    async def update_job_status(
        self,
        job_id: str,
        status: str,
        worker_id: str = "",
        result: str = "",
    ) -> None:
        mapping = {
            "status": status,
            "updated_at": str(time.time()),
        }
        if worker_id:
            mapping["worker_id"] = worker_id
        if result:
            mapping["result"] = result
        await self.redis.hset(f"job:{job_id}", mapping=mapping)
